# 上机实验12：贝叶斯分类

## 任务1：贝叶斯定理

给定如下的队伍信息
- 队0获胜的概率prob_win_0 = 0.65
- 队1获胜时队1是主场的概率prob_win_1_home_1 = 0.75
- 队0获胜时队1是主场的概率prob_win_0_home_1 = 0.3
>  请计算：队1在客场取胜的概率p(Y=1|X=0)，即prob_win_1_home_0，小数点保留三位。

> 待完善代码

In [1]:
## 变量默认值代码
prob_win_0 = 0.65
prob_win_1_home_1 = 0.75
prob_win_0_home_1 = 0.3

## 计算队1获胜的概率prob_win_1
prob_win_1 = 1 - prob_win_0
## 计算队1取胜时队0是东道主的概率
prob_win_1_home_0 = 1 - prob_win_1_home_1

## 计算队0取胜时队0是东道主的概率
prob_win_0_home_0 = 1 - prob_win_0_home_1

## 计算队0是东道主的概率
prob_home_0 = prob_win_1 * prob_win_1_home_0 + prob_win_0 * prob_win_0_home_0  # 全概率公式
    
## 队1在客场取胜的概率
prob_win_1_home_0 = prob_win_1 * prob_win_1_home_0 / prob_home_0 #联合概率除以客场的概率
print(prob_win_1_home_0)

0.16129032258064516


> 预期输出

0.161

## 使用朴素贝叶斯进行文本分类

下面三个任务是使用朴素贝叶斯分类器进行垃圾邮件的预测，在此之前需要了解使用朴素贝叶斯进行文本分类要考虑的三个问题：

![](https://ai-studio-static-online.cdn.bcebos.com/78572efcb8cf4f06b760c8dc7c4bb62327030b4f4fc5487bb324964916e8da30)


## 任务2：邮件文本向量化

给定经过分词处理后的正常邮件和垃圾邮件各25封，对邮件内容分词后形成了词库列表vocab，存储到vocab_list.txt中。
- 请自定义函数createVec(vocab, mail_text)，输入词库列表vocab和单个分词后的邮件内容列表mail_text，输出参数为向量化文本列表mail_vec和邮件单词在词库列表vocab中的位置列表matchIndex。该自定义函数的作用是，将mail_text向量化，维度为vocab中的单词总和。比如，mail_text中含有vocab中的单词b，mail_vec的维度b的取值即为1，而matchIndex中记录单词b在vocab中的位置。
- 给定测试邮件内容列表test_mail，使用自定义函数createVec(vocab, mail_text)返回向量化文本列表test_mail_vec。

> 给定数据的格式如下：vocab = ['am', 'hacker', 'machine',...]

> 待补全代码

In [4]:
import pandas as pd
import numpy as np
from numpy import linalg as LA

# 读取数据
vocab = pd.read_csv('vocab_list.txt', header=None).values.tolist()
li = []
for i in vocab:
    li.append(i[0])
vocab = li

test_mail = ['hi','peter','jose','town','want','meet','keep','things','going','interesting','stuff','let','know','eugene']


def createVector(vocab, mail_text): 
    
    ## 获取vocab的长度
    len_vocab = len(vocab)
    
    ## 零值初始化向量，长度为len_vocab
    mail_vec = np.zeros(len_vocab)
    
    ## 定义一个用来存储匹配单词位置的变量
    matchIndex = []

    ## 遍历mail_text中的元素
    for value in mail_text:
        
        ## 判断value在vocab中的位置，将mail_vec对应位置上的元素值加1
        if value in vocab:
            
            ## 获取value在vocab中的索引value_index
            value_index = vocab.index(value)
            
            ## mail_vec中value_index位置上的元素赋值为1
            mail_vec[value_index] = 1

            ## matchIndex记录下匹配单词的位置索引index 
            matchIndex.append(value_index)
        else:
            continue
    
    ## 返回结果
    return mail_vec, matchIndex
## 获取test_mail_vec
test_mail_vec, matchIndex = createVector(vocab, test_mail)
print('该邮件内容的向量化文本为')
print(test_mail_vec)
print('该邮件内容在词库列表vocab中的位置索引向量为')
print(matchIndex)

该邮件内容的向量化文本为
[0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0.
 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1. 0.
 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 1. 0. 0.
 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0. 0

> 预期输出

![](https://ai-studio-static-online.cdn.bcebos.com/814761ab3ea9455ba6908f62ec927e6d22715d4ddd8f4e59894de13a9fe4ced2)


## 任务3：朴素贝叶斯分类器的参数估计

从50封邮件中选取各20封正常邮件和垃圾邮件，经过分词处理后，组成邮件训练集列表train_data。已知训练集的标签列表为train_label，词库列表为vocab，令垃圾邮件标记为1，正常邮件标记为0：
- 使用createVector函数将train_data中的记录向量化，生成train_matrix；
- 自定义函数createNBClassifier(train_matrix, train_label)函数，输出判定所需要的p(xi|c)，以及p(c)；

给定数据的格式如下：

> train_data = [['I', 'am', 'data', 'hacker'...], ['data', 'mining','machine'],...]
> 
> train_label = [1,0,1,1,...]
> 
> vocab = ['am', 'hacker', 'machine',...]

> 待补全代码

In [12]:
import pandas as pd
import numpy as np
from numpy import linalg as LA

# 数据处理与读取
train_data = pd.read_csv('train_data.txt', header=None).values.tolist()
#print(train_data)
li = []
for i in train_data:
    li.append(i[0].split(" "))
train_data = li

train_label = pd.read_csv('train_label.txt', header=None).values.tolist()
li = []
for i in train_label:
    li.append(i[0])
train_label = li


# 使用createVector函数将train_data向量化，结果保存在train_matrix
# 请在下方完成代码
train_matrix = []
train_matchIndex = []
for sample in train_data: #######
    sample_matrix, sample_matchIndex = createVector(vocab, sample)
    train_matrix.append(sample_matrix)
    train_matchIndex.append(sample_matchIndex)



def createNBClassifier(train_matrix, train_label):
    ## 获取训练集的数量num
    num = len(train_matrix)

    ## 获取train_matrix的维度num_col
    num_col = len(train_matrix[0])

    ## 统计c=1下x_{i}出现的频次列表p1_num，形如[num(x_1|c=1), num(x_2|c=1),...]; 同理 c=0
    ## 为了防止出现概率为零的情况，我们将所有词出现的频次初始化为1
    p1_num = np.ones(num_col)
    p0_num = np.ones(num_col)

    ## 垃圾邮件和正常邮件分别出现的单词的数量，初始化为0.0  
    p1_vec_denom = 0.
    p0_vec_denom = 0.

    for index in range(num):
        if train_label[index] == 1:
            ## 单词出现的位置值为1，不出现值为0，因此可以直接相加，累计词频，train_matrix的元素为list，需要转换为ndarray，便于向量直接相加
            p1_num = p1_num + np.array(train_matrix[index])
            ## 统计垃圾邮件中的单词数
            p1_vec_denom = p1_vec_denom + np.array(train_matrix[index]).sum()  
        else:
            ## 单词出现的位置值为1，不出现值为0，因此可以直接相加，累计词频，train_matrix的元素为list，需要转换为ndarray，便于向量直接相加
            p0_num = p0_num + np.array(train_matrix[index])
            ## 统计正常邮件中的单词数
            p0_vec_denom = p0_vec_denom + np.array(train_matrix[index]).sum()

    ## 根据从训练集统计的频次，计算[p(x_i|c=1), p(x_2|c=1)...] [p(x_i)|c=0), p(x_i|c=0),...]
    ## 为了防止向下溢出，概率取对数；为了配合初始化值为1的频次，分母+2
    p1_vec = np.log(p1_num / (p1_vec_denom + 2))
    p0_vec = np.log(p0_num / (p0_vec_denom + 2))

    ## 计算垃圾邮件出现的概率
    pc1 = (np.array(train_label) == 1).sum() / num

    ## 计算正常邮件出现的概率
    pc0 = (np.array(train_label) == 0).sum() / num

    ## 返回结果
    return p1_vec, p0_vec, pc1, pc0

## 使用自定义函数createNBClassifier返回结果p1_vec, p0_vec, pc1, pc0
p1_vec, p0_vec, pc1, pc0 = createNBClassifier(train_matrix, train_label)

print('pc1=',pc1)
print('pc0=',pc0)
print('p1_vec')
print(p1_vec)
print('p0_vec')
print(p0_vec)

pc1= 0.5
pc0= 0.5
p1_vec
[-4.70411013 -5.21493576 -4.70411013 -4.70411013 -4.11632347 -6.31354805
 -5.62040087 -5.62040087 -6.31354805 -5.62040087 -4.52178858 -5.21493576
 -5.21493576 -5.62040087 -6.31354805 -5.62040087 -5.62040087 -5.21493576
 -6.31354805 -4.92725369 -6.31354805 -5.62040087 -5.21493576 -5.62040087
 -5.62040087 -5.62040087 -5.62040087 -5.62040087 -6.31354805 -5.21493576
 -5.21493576 -5.62040087 -5.21493576 -5.62040087 -5.62040087 -5.21493576
 -4.3676379  -3.67449072 -6.31354805 -4.92725369 -5.62040087 -5.62040087
 -5.21493576 -5.21493576 -5.62040087 -5.21493576 -5.62040087 -5.62040087
 -4.3676379  -5.62040087 -5.62040087 -4.70411013 -4.70411013 -6.31354805
 -5.62040087 -5.21493576 -5.21493576 -5.62040087 -5.62040087 -5.62040087
 -4.3676379  -5.62040087 -6.31354805 -5.21493576 -5.21493576 -4.70411013
 -5.62040087 -6.31354805 -4.92725369 -6.31354805 -6.31354805 -5.62040087
 -4.2341065  -6.31354805 -4.92725369 -5.21493576 -6.31354805 -5.62040087
 -6.31354805 -6.31354805 -

> 预期输出（部分截图）

![](https://ai-studio-static-online.cdn.bcebos.com/48df39c8b80e4d33a61dd160efcf82e34ea06da2c0c34488ac3e2ed1ddf12558)


## 任务4：贝叶斯估计用于分类预测

将剩余的10封邮件内容进行文本向量化，组成测试集邮件内容列表test_matrix。给定训练集拟合生成的朴素贝叶斯分类模型，返回预测测试集所需要的概率p1_vec，p0_vec，pc1和pc0，以及测试集的真实标签列表test_label。
- 自定义函数predict(test_data, p1_vec, p0_vec, pc1, pc0)函数，判断输入的邮件是否为垃圾邮件；
- 使用自定义的预测函数predict()，对test_data进行预测，输出预测标签列表pred_label；
- 根据pred_label和test_label计算朴素贝叶斯模型的错误率error_rate。

给定的数据格式如下：

> test_data = [[0,0,0,1,0,0,...], [1,0,0,0,1,0,...]]
> 
> test_label = [0,0,1,1,...]
> 
> p1_vec = np.array([-0.02, -0.003, ...])
> 
> p0_vec = np.array([-0.02, -0.003, ...])
> 
> pc1 = 0.5
> 
> pc0 = 0.5

> 前序依赖数据
- 用于向量化邮件的词典数据vocab
- 前序学习出的类条件分布概率 p1_vec，p0_vec，及类别分布先验 pc1，pc0

> 前序代码操作步骤：
- 运行任务2中的自定义函数 createVector()
- 运行任务3中的自定义函数 createNBClassifier()，获得类条件分布概率 p1_vec，p0_vec，及类别分布先验 pc1，pc0


> 待补全代码

In [13]:
## 加载本次实验测试数据
test_label = pd.read_csv('test_label.txt', header=None).values.tolist()
li = []
for i in test_label:
    li.append(i[0])
test_label = li

test_data = pd.read_csv('test_data.txt', header=None).values.tolist()
li = []
for i in test_data:
    li.append(i[0].split(" "))
test_data = li
 
# 使用createVector函数将test_data向量化，结果保存在test_matrix
test_matrix = [createVector(vocab, sample)[0] for sample in test_data]###
# 请在下方完成代码

def predict(test_matrix, p1_vec, p0_vec, pc1, pc0):

    ## 声明预测标签列表，初始值为0
    pred_label = np.zeros(len(test_matrix))
    
    ## 遍历测试机中的每一个样本
    for index, record in enumerate(test_matrix):
        ## 将list转换为ndarray
        record_array = np.array(record)
        
        ## 根据公式 log(p(c|x))正比于 log(p(x_1|c)p(x_2|c)...p(x_k|c)p(c))
        ## 计算测试样本属于垃圾邮件的概率
        p1 = p1_vec[record == 1].sum() + np.log(pc1)
    
        ## 计算测试样本属于正常邮件的概率
        p0 = p0_vec[record == 1].sum() + np.log(pc0)
        ## 确认预测标签
        if p1 > p0:
            ##将索引为index的测试标签值改为1
            pred_label[index] = 1
        else:
            continue
    
    ## 返回结果
    return pred_label

## 使用自定义函数predict预测测试集的标签
pred_label = predict(test_matrix, p1_vec, p0_vec, pc1, pc0)
## 获取预测标签与真实标签的差，diff类型为ndarray
diff = np.array(np.abs(pred_label - test_label))

## 计算错误率
error_rate = np.abs(diff).sum() / len(test_label)

## 打印错误率
print('预测错误率为')
print(round(error_rate, 2))
print('真实标签为')
print(test_label)
print('预测标签为')
print(pred_label)
print('预测标签是否不一致向量')
print(diff)

预测错误率为
0.1
真实标签为
[0, 0, 0, 0, 0, 1, 1, 1, 1, 1]
预测标签为
[0. 0. 0. 1. 0. 1. 1. 1. 1. 1.]
预测标签是否不一致向量
[0. 0. 0. 1. 0. 0. 0. 0. 0. 0.]


> 预期输出

![](https://ai-studio-static-online.cdn.bcebos.com/d8343d3ae7a241d1ab842070c121f5ddc941362637e24512967d1edfc766e8a3)
